flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputSelectableProcessor.java

```diff
@@ -0,0 +1,433 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.SimpleCounter;
+import org.apache.flink.runtime.io.disk.iomanager.IOManager;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.metrics.groups.OperatorMetricGroup;
+import org.apache.flink.streaming.api.operators.InputSelectable;
+import org.apache.flink.streaming.api.operators.InputSelection;
+import org.apache.flink.streaming.api.operators.TwoInputStreamOperator;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.metrics.WatermarkGauge;
+import org.apache.flink.streaming.runtime.streamrecord.StreamElement;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.streamstatus.StreamStatus;
+import org.apache.flink.streaming.runtime.streamstatus.StreamStatusMaintainer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Input reader for {@link org.apache.flink.streaming.runtime.tasks.TwoInputSelectableStreamTask}
+ * in the case that the operator is InputSelectable.
+ *
+ * @param <IN1> The type of the records that arrive on the first input
+ * @param <IN2> The type of the records that arrive on the second input
+ */
+@Internal
+public class StreamTwoInputSelectableProcessor<IN1, IN2> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(StreamTwoInputSelectableProcessor.class);
+
+	private volatile boolean continuousProcessing = true;
```

Comment:
For input processing , using the internal loop to continuously process records is about 1000 ops/ms faster than using the external loop. `volatile boolean continuousProcessing` is introduced to interrupt the input processing when the task is canceled or some cases are need to interrupt in the future. For production code, it is assigned `false` only in `StreamTwoInputSelectableProcessor#stop ()` now.

The code structures for the internal and external loops are as follows.

1. Uses a internal loop in the input processor to continuously process records
```
=== TwoInputXStreamTask.java ===
@Override
protected void run() throws Exception {
	// cache processor reference on the stack, to make the code more JIT friendly
	final StreamTwoInputXProcessor<IN1, IN2> inputProcessor = this.inputProcessor;
	...
	while (running && inputProcessor.processInput()) {
		// all the work happens in the "processInput" method
	}
	...
}

@Override
protected void cancelTask() {
	running = false;
	if (inputProcessor != null) {
		inputProcessor.stop();
	}
}

=== StreamTwoInputXProcessor.java ===
public boolean processInput() throws Exception {
	while (continuousProcessing) {
		...
		if (element.isRecord()) {
			continue;
		}
		...
	}
	return true;
}

public void stop() {
	continuousProcessing = false;
	...
}
```


2. Uses a external loop in the stream task to continuously process records
```
=== TwoInputXStreamTask.java ===

@Override
protected void run() throws Exception {
	// cache processor reference on the stack, to make the code more JIT friendly
	final StreamTwoInputXProcessor<IN1, IN2> inputProcessor = this.inputProcessor;
	...
	while (running && inputProcessor.processInput()) {
		// all the work happens in the "processInput" method
	}
	...
}

@Override
protected void cancelTask() {
	running = false;
}

=== StreamTwoInputXProcessor.java ===
public boolean processInput() throws Exception {
	while (true) {
		...
		if (element.isRecord()) {
			return true;
		}
		...
	}
}
```

Reply:
> For input processing , using the internal loop to continuously process records is about 800~1000 ops/ms faster than using the external loop

Hmmm, it would be strange. I can see only two possibilities for this outcome:

1. You accidentally measured the impact of adding second `volatile` and this was the actual performance regression
2. JIT was not able to inline `StreamTwoInputSelectableProcessor#processInput` since it's too big.

Regarding point 2. - I wasn't commenting on it yet, since I was doing only a high level review, but before merging this code, a lot of those methods should be broken down to smaller ones. So if you will be sure that 1. is not the case, please refactor the code of `StreamTwoInputSelectableProcessor#processInput` or `NetworkInput#pollNextElement` by extracting some of the code to a smaller methods. For example both branches of the `if (readingInputIndex == 0) {` could be extracted.

My point is that keeping those two loops working both of them on `while (continuousProcessing)` condition doesn't make sense. One of them should be removed. It would be easier to keep the things how they were, since this pattern of loop inside `XInputStreamTask` and `XInputStreamProcessor` is also in `OneInput` case. If for whatever reasons we decide to change it for `TwoInput`, we should make sure that `OneInput` is kept up to date and in sync.

Reply:
> You accidentally measured the impact of adding second volatile and this was the actual performance regression

I also tested `StreamTwoInputProcessor`, and it is the same result.

Reply:
In addition, when the test uses an external loop to process records continuously, as shown in the code, `while (continuousProcessing)` is changed to `while (true)`.

Reply:
> My point is that keeping those two loops working both of them on while (continuousProcessing) condition doesn't make sense. One of them should be removed.

I prefer to keep the internal loop because it is more JIT-friendly, and we do not need to worry about how to make that `processInput()` is inline-optimized.

Reply:
> I prefer to keep the internal loop 

If you have strong preference here, that's fine with me. In that case please adjust the existing processors in a separate commit and drop the outer loop.

Just to be clear, this `StreamTwoInputSelectableProcessor#processInput()` should be split into smaller methods before merging this PR regardless of the outer/inner loop discussion.

Reply:
> Just to be clear, this StreamTwoInputSelectableProcessor#processInput() should be split into smaller methods before merging this PR regardless of the outer/inner loop discussion.

I agree, and splitting after the performance problem has been solved which reduce the interference to performance tuning.

Reply:
> If you have strong preference here, that's fine with me. In that case please adjust the existing processors in a separate commit and drop the outer loop.

Okay, I will create a separate JIRA to drop the outer loop in   `OneInputStreamTask#run()` and `TwoInputStreamTask#run()`.